File: Funcional/keys_genarator.hs
module KeysGenerator (
    calculaN,
    getE,
    getD,
    generateKeys
)

import System.Random

calculaN :: Int -> Int -> Int
calculaN p q = p * q

totiente :: Int -> Int -> Int
totiente p q = (p-1) * (q-1) 

newRand = randomIO :: IO Int
 
getE :: Int -> IO()
getE tot = 
    do 
        f1 <- newRand
        let aux = (abs(f1)) `mod` tot + 2
        if (gcd tot aux) == 1 
            then
                do
                    putStr ("\n  e: " ++ (show aux) ++ "\n")
                    getD tot aux 0
                    else getE tot

getD :: Int -> Int -> Int -> IO()
getD to e aux = 
    if ((e * aux) `mod` to) == 1
        then putStr ("\nChave Privada: \n  d: "++ show (aux))
        else getD to e (aux + 1)

generateKeys :: Int -> Int -> IO()
generateKeys p q = do
    putStr ("\n\nChave pública: \n  n: "++show(n))
    getE t   
    putStr ("\n  Primo 1: "++show(p)++"\n  Primo 2: "++show(q)++ "\n\n")
    where   
        n = calculaN p q
        t = totiente p q

File: Logico/keysGenarator.pl
calculaN(P,Q,R):- R is P*Q.

totiente(P,Q,R):- R is (P-1)*(Q-1).

getE(Totiente,Return):- 
        random(1,1000000,Ram),
        Aux is Ram mod (Totiente + 2),
        Result is gcd(Totiente,Aux),
        Result =:= 1 -> Return is Aux;
        getE(Totiente,Return).

getD(Totiente, E, Aux, Return):-
        (E * Aux) mod Totiente =:= 1 -> Return is Aux;
        Aux2 is Aux + 1,
        getD(Totiente, E, Aux2, Return).

input(Number):-
    read_line_to_codes(user_input, Codes),
    string_to_atom(Codes, Atom),
    atom_number(Atom, Number).

geraKeys(P,Q):- 
        calculaN(P,Q,N),
        totiente(P,Q, Tot),
        getE(Tot,E),
        getD(Tot,E,0,D),
        nl,
        write("\nChave pública: \n  n: "), writeln(N),
        write("  e: "),writeln(E),
        write("\nChave Privada: \n  d: "), writeln(D),
        write("  Primo 1: "), writeln(P), write("  Primo 2: "), writeln(Q).